tracing: Capture events in tracing data

rnburn opened the pull request:
Signed-off-by: Ryan Burn <ryan.burn@gmail.com>

*Description*: Extents HttpTracer to include a log method for recording metadata with an associated timestamp.

Adds code to attach logs for the events in StreamInfo.

*Risk Level*: Moderate
*Testing*: Added unit test coverage.

objectiser commented on source/extensions/tracers/zipkin/zipkin_tracer_impl.h:
```diff
@@ -62,6 +62,11 @@ class ZipkinSpan : public Tracing::Span {
+  /**
+   * Do nothing for now. Log records could potentially be translated to tags in the future.
+   */
+  void log(SystemTime /*timestamp*/, const std::string& /*event*/) override {}
```
In zipkin, this would be represented by an annotation, e.g. https://github.com/envoyproxy/envoy/blob/master/source/extensions/tracers/zipkin/tracer.cc#L22 but the annotation value would be the `event`. Would be good if you could add this impl as part of the PR.

mattklein123 replied:
+1

mattklein123 commented:
> LGTM - although my only concern would be whether this information should be reported by default - maybe some configuration property should be used to enable more detailed tracing?

I have the same general feeling. Also, should this be documented somewhere about standard logs that we emit if configured?

/wait

objectiser commented:
@rnburn Thanks for adding the support in zipkin tracer.

However I don't think the flag should be enforced in the log method - as this means all logs would be discarded. It would be better if the flag is used to guard the code that is reporting the detailed information (https://github.com/envoyproxy/envoy/pull/5915/files#diff-84a39b5a7754b23d0737abed8afdea43R128 in this case).

This would mean that other proxy code still has the option to report log events that are not considered detailed info - and also some tags could also be guarded by this flag, so some tags would only be reported if detailed info is required. So actually might be better to make the flag name more general (not logging or stream event specific).

@mattklein123 Thoughts? Also will be on PTO from tomorrow until 1st March so someone else may need to continue with review.

rnburn commented:
> However I don't think the flag should be enforced in the log method - as this means all logs would be discarded. It would be better if the flag is used to guard the code that is reporting the detailed information (https://github.com/envoyproxy/envoy/pull/5915/files#diff-84a39b5a7754b23d0737abed8afdea43R128 in this case).

That's the only code that does logging. 

Did you mean to have it specified via the `Tracing::Config` classes? -- I'm not sure if that would be a good place for the option since those classes aren't set by the yaml configuration.

objectiser commented:
> That's the only code that does logging.

That is the only code that current does logging - but other code/extensions may start using it now that it is part of the tracer API.

> Did you mean to have it specified via the Tracing::Config classes? -- I'm not sure if that would be a good place for the option since those classes aren't set by the yaml configuration.

No, just mean that the logic for handling log events should not be pushed into the tracer implementations.

mattklein123 commented on api/envoy/config/trace/v2/trace.proto:
```diff
@@ -47,7 +47,11 @@ message Tracing {
     }
+
+    // Whether to attach logs for streaming events to spans.
+    bool log_stream_events = 4;
```
Can you add a release note for this? Do we need other docs as well for that this does exactly (in terms of what events are logged by default, etc.)? I have a general question as to whether this should be a config option on the HTTP connection manager vs. the driver, and gated there. I think that makes more sense? WDYT?

mattklein123 commented on include/envoy/server/tracer_config.h:
```diff
@@ -19,15 +19,14 @@ class TracerFactory {
   /**
-   * Create a particular HttpTracer implementation. If the implementation is unable to produce an
-   * HttpTracer with the provided parameters, it should throw an EnvoyException in the case of
+   * Create a particular Driver implementation. If the implementation is unable to produce an
```
nit: s/an/a

mattklein123 commented on include/envoy/server/tracer_config.h:
```diff
@@ -19,15 +19,14 @@ class TracerFactory {
-   * HttpTracer with the provided parameters, it should throw an EnvoyException in the case of
+   * Create a particular Driver implementation. If the implementation is unable to produce an
+   * Driver with the provided parameters, it should throw an EnvoyException in the case of
    * general error or a Json::Exception if the json configuration is erroneous. The returned
```
I think the JSON part of this comment no longer applies, can you clean it up?

rnburn commented:
sill working on this

rnburn commented:
I move the option to ConnectionManager and renamed it to `verbose` so that it can be also used for controlling other types of annotations. And I added a release note.
